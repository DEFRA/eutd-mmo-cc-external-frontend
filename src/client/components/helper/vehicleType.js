import _ from 'lodash';

export const PLANE_KEY                   = 'plane';
export const TRUCK_KEY                   = 'truck';
export const TRAIN_KEY                   = 'train';
export const CONTAINER_VESSEL_KEY        = 'containerVessel';
export const DIRECT_LANDING_KEY          = 'directLanding';
export const DESTINATION_KEY             = 'departurePlace';
export const RAILWAY_BILL_NUMBER_KEY     = 'railwayBillNumber';
export const FLIGHT_NUMBER_KEY           = 'flightNumber';
export const CONTAINER_NUMBER_KEY        = 'containerNumber';
export const FLAG_STATE_KEY              = 'flagState';
export const VESSEL_NAME_KEY             = 'vesselName';
export const REGISTRATION_NUMBER_KEY     = 'registrationNumber';
export const NATIONALITY_OF_VEHICLE_KEY  = 'nationalityOfVehicle';
export const TRANSPORT_TYPE_KEY          = 'vehicle';
export const TRANSPORT_SELECTION_KEY     = 'transportSelection';

export const ALL_SUPPORTED_VEHICLE_TYPES = [
  TRUCK_KEY,
  PLANE_KEY,
  TRAIN_KEY,
  CONTAINER_VESSEL_KEY,
  DIRECT_LANDING_KEY
];

export const ALL_SUPPORTED_VEHICLE_TYPES_NO_DIRECT_LANDING = [
  TRUCK_KEY,
  PLANE_KEY,
  TRAIN_KEY,
  CONTAINER_VESSEL_KEY
];

export const capitalizeFirstLetter = word => {
  return word.toLowerCase().split(' ').map(_.capitalize).join(' ');
};

export const VEHICLE_TYPE_FIELDS = {
  [PLANE_KEY]: [DESTINATION_KEY, TRANSPORT_TYPE_KEY, FLIGHT_NUMBER_KEY, CONTAINER_NUMBER_KEY],
  [TRUCK_KEY]: [DESTINATION_KEY, TRANSPORT_TYPE_KEY, REGISTRATION_NUMBER_KEY, NATIONALITY_OF_VEHICLE_KEY],
  [TRAIN_KEY]: [DESTINATION_KEY, TRANSPORT_TYPE_KEY, RAILWAY_BILL_NUMBER_KEY],
  [CONTAINER_VESSEL_KEY]: [DESTINATION_KEY, TRANSPORT_TYPE_KEY, VESSEL_NAME_KEY, FLAG_STATE_KEY, CONTAINER_NUMBER_KEY],
  [DIRECT_LANDING_KEY]: []
};
