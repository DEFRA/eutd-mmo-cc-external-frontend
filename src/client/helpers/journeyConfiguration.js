export const catchCertificateJourney = 'catchCertificate';
export const storageNoteJourney = 'storageNotes';
export const processingStatementJourney = 'processingStatement';

export const catchCertificatesTitle = 'ccCommonTitle';
export const storageDocumentTitle = 'sdCommonTitle';
export const processingStatementTitle = 'psCommonTitle';

export const catchCertificateJourneyText = 'catchCertificate';
export const processingStatementJourneyText = 'processingStatement';
export const storageDocumentJourneyText = 'storageNotes';

export const exporterAddressLookupTitle = 'commonWhatExportersAddressWhatIsExportersAddress';
export const processingPlantLookupTitle = 'psProcessingPlantAddressWhatIsProcessingPlantAddress';
export const storageDocumentLookupTitle = 'sdWhatStorageFacilityAddressHeaderTitle';