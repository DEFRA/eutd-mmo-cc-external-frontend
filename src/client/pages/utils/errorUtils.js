import i18n from '../../../i18n';
import enTranslations from '../../../locales/uk/en.json';

export function toGovukErrors(obj) {
  return Object.keys( obj).map( o => {
    return {targetName: o, text: obj[o], key: o, message: (obj[o] in enTranslations) ? i18n.t(obj[o]) : obj[o]};
  });
}

export function onHandleErrorClick(targetName, speciesRow, vesselIndex) {
  let elem;

  if (speciesRow !== undefined && vesselIndex !== undefined) {
    elem = document.getElementById(targetName + '-' + speciesRow + '-' + vesselIndex);
  } else {
    elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  }

  if (elem) {
    elem.scrollIntoView({ behavior: 'smooth' });
    elem.focus({ preventScroll: true});
  }
}

export function onHandleErrorClickWithoutFocus(targetName) {
  const elem = document.getElementsByName(targetName)[0] || document.getElementById(targetName);
  elem.scrollIntoView({ behavior: 'smooth' });
}

export function scrollToErrorIsland() {
  setTimeout( () => {
    document.getElementById('errorIsland').scrollIntoView({behavior: 'smooth'});
    document.getElementById('errorIsland').focus({ preventScroll: true});
  },100);
}

export function scrollToField(id) {
  setTimeout( () => {
    document.getElementById(id).scrollIntoView({behavior: 'smooth'});
    document.getElementById(id).focus({ preventScroll: true});
  },100);
}

export function scrollToFieldName(name) {
  setTimeout( () => {
    document.getElementsByName(name)[0].scrollIntoView({behavior: 'smooth'});
    document.getElementsByName(name)[0].focus({ preventScroll: true});
  },100);
}

