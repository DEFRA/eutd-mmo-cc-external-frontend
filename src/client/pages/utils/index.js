import React, {Fragment} from 'react';

export * from './errorUtils';


export function formatAddress( ...adresssLines) {
  return adresssLines.filter( a => a ).map( (a, index) => <Fragment key={index}>{a}<br/></Fragment> );
}

export function constructAddress(addressArray) {
  let address = '';
  if (addressArray.length === 4) {
    if (addressArray[0]) {
      address += addressArray[0];
    }
    if (addressArray[0] && addressArray[1]) {
      address += ' ';
    }
    if (addressArray[1]) {
      address += addressArray[1];
    }
    if (addressArray[2]) {
      address += ', ' + addressArray[2];
    }
    if (addressArray[3]) {
      address += '. ' + addressArray[3];
    }
  }
  return address;
}

export const hasAdminOverride = (exportPayload) => {
  const isLandingOverridden = (landing) =>
    landing &&
    landing.model &&
    landing.model.vessel &&
    landing.model.vessel.vesselOverriddenByAdmin;

  return exportPayload &&
    exportPayload.items &&
    exportPayload.items.some(
      (item) =>
        item.landings &&
        item.landings.some((landing) => isLandingOverridden(landing))
    )
    ? true
    : false;
};