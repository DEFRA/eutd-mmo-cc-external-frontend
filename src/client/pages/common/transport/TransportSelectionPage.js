import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import _ from 'lodash';

import {
  Main,
  Header,
  MultiChoice,
  GridRow,
  GridCol
} from 'govuk-react';

import SelectRadio from '../../../components/elements/SelectRadio';

import PageTitle from '../../../components/PageTitle';
import {
  addTransport,
  getTransportDetails,
  clearTransportDetails,
  saveTransport,
  dispatchApiCallFailed,
  dispatchClearErrors,
  getExportCountry,
  getStorageNotesFromRedis
} from '../../../actions';

import {
  getLandingType
} from '../../../actions/landingsType.actions';

import { camelCaseToSpacedUpperCase } from '../../../helpers/string';
import { forwardUri } from '../../../helpers/vehicleRouteLookup';
import BackLinkWithErrorClearOut from '../../../components/elements/BackLinkWithErrorClearOut';
import { scrollToErrorIsland, onHandleErrorClick } from '../../utils';
import PageTemplateWrapper from '../../../components/PageTemplateWrapper';
import HelpLink from '../../../components/HelpLink';
import SaveAsDraftButton from '../../../components/SaveAsDraftButton';
import ContinueButton from '../../../components/elements/ContinueButton';
import ErrorIsland from '../../../components/elements/ErrorIsland';
import { catchCertificateJourney } from '../../../helpers/journeyConfiguration';
import displayProgressPage from './helpers/displayProgressPage';
import BackToProgressLink from '../../../components/BackToProgressLink';
import { withTranslation } from 'react-i18next';
import { t } from 'i18next';

class TransportSelectionPage extends Component {
  constructor(props) {
    super(props);
  }

  onChange = e => {
    this.props.clearTransportDetails();
    this.props.addTransport({ [e.target.name]: e.target.value });
  };

  onSubmit = async e => {
    e.preventDefault();

    try {
      const { path, journey } = this.props.route;
      const documentNumber = this.props.match.params['documentNumber'];
      const transport = await this.props.saveTransport(path, journey, false, documentNumber);

      // TODO - need to investigate dynamic routing here
      let to = forwardUri(transport.vehicle, this.props.route);

      this.props.history.push(to.replace(':documentNumber', documentNumber));
    } catch (error) {
      scrollToErrorIsland();
    }
  };

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    try {
      const { path, journey, saveAsDraftUri } = this.props.route;
      const documentNumber = this.props.match.params['documentNumber'];
      const isTransportSavedAsDraft = true;
      await this.props.saveTransport(path, journey, isTransportSavedAsDraft, documentNumber);
      this.props.history.push(saveAsDraftUri);
    } catch (error) {
      scrollToErrorIsland();
    }
  };

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  displayLandingEntryOptions() {
    const documentNumber = this.props.match.params['documentNumber'];
    const { landingsEntryUri } = this.props.route;
    this.props.history.push(landingsEntryUri.replace(':documentNumber', documentNumber));
  }

  hasLandingEntryOption() {
    const { getAddedLandingsType } = this.props;
    return getAddedLandingsType && getAddedLandingsType.landingsEntryOption !== null && getAddedLandingsType.generatedByContent === false;
  }

  componentDidMount = async () => {
    const { journey } = this.props.route;
    const documentNumber = this.props.match.params['documentNumber'];

    window.scrollTo(0, 0);
    try {
      await this.props.getTransportDetails(this.props.route.journey, documentNumber);
      await this.props.getExportCountry(documentNumber);
      await this.props.getFromRedis(documentNumber);

      if (journey === catchCertificateJourney) {
        await this.props.getLandingType(documentNumber);

        const isDirectLanding = this.props.getAddedLandingsType || {};

        if (isDirectLanding.landingsEntryOption === 'directLanding') {
          displayProgressPage(this.props);
          return;
        }

        if (!this.hasLandingEntryOption()) {
          this.displayLandingEntryOptions();
          return;
        }
      }

    } catch (e) {
      console.error(e);
    }
  };

  componentWillUnmount() {
    this.props.clearTransportDetails();
  }

  componentDidUpdate = () => {
    const { transport } = this.props;

    if (transport.unauthorised === true) {
      this.unauthorised();
    }


  };

  renderTransportOptions = () => {
    return this.props.route.vehicleTypes.map(type => {
      const typeInCaps = camelCaseToSpacedUpperCase(type);
      const label = {
        'Direct landing': 'Fishing vessel'
      }[typeInCaps] || typeInCaps;

      if (
        this.props.transport.vehicle &&
        type.toLowerCase() === this.props.transport.vehicle.toLowerCase()
      ) {
        return (
          <SelectRadio
            key={'vehicleType' + type}
            id={type}
            value={type}
            name="vehicle"
            defaultChecked
          >
            {t(label)}
          </SelectRadio>
        );
      }
      return (
        <SelectRadio
          key={'vehicleType' + type}
          id={type}
          value={type}
          name="vehicle"
        >
          {t(label)}
        </SelectRadio>
      );
    });
  };

  render = () => {
    const { vehicleError, errors } = this.props.errors;
    const { dispatchClearErrors: _dispatchClearErrors, history } = this.props;

    const saveAsDraftFormActionUrl = '/orchestration/api/v1/transport/add/saveAsDraft';

    const {
      title,
      previousUri,
      truckCmrUri,
      planeDetailsUri,
      trainDetailsUri,
      containerVesselDetailsUri,
      summaryUri,
      progressUri,
      journey,
      path,
      saveAsDraftUri
    } = this.props.route;

    const documentNumber = (this.props.match && this.props.match.params)
      ? this.props.match.params['documentNumber']
      : '';

    return (
      <Main>
        <PageTitle title={`${!_.isEmpty(errors) ? 'Error: ' : ''}${t('commonTransportSelectionPageTitle')} - ${t(title)}`} />
        {errors && (
          <ErrorIsland
            errors={errors.map(err => ({
              message: t(err.text),
              key: err.targetName
            }))}
            onHandleErrorClick={onHandleErrorClick}
          />
        )}
        <GridRow>
          <GridCol>
            <BackLinkWithErrorClearOut
              backUri={previousUri.replace(':documentNumber', documentNumber)}
              clearErrors={_dispatchClearErrors}
              history={history}
              labeltext={t('commonBackLinkBackButtonLabel')}
            />
            <Header>
              {t('commonTransportSelectionPageTitle')}
            </Header>
          </GridCol>
        </GridRow>

        <form
          action="/orchestration/api/v1/transport/add"
          method="POST"
          onSubmit={this.onSubmit}
        >
          <input type="hidden" name="currentUri" value={path} />
          <input type="hidden" name="truckCmrUri" value={truckCmrUri} />
          <input type="hidden" name="planeDetailsUri" value={planeDetailsUri} />
          <input type="hidden" name="trainDetailsUri" value={trainDetailsUri} />
          <input
            type="hidden"
            name="containerVesselDetailsUri"
            value={containerVesselDetailsUri}
          />
          <input type="hidden" name="summaryUri" value={summaryUri} />
          <input type="hidden" name="journey" value={journey} />
          <input type="hidden" name="dashboardUri" value={saveAsDraftUri} />

          <GridRow>
            <GridCol>
              <div id="vehicle">
                <MultiChoice label={t('commonTransportSelectionSelectTypeTransportLabel')}
                  meta={{ touched: vehicleError, error: t(vehicleError) }}
                  onChange={this.onChange}
                >
                  <fieldset>
                    <legend className="visually-hidden">{t('commonTransportSelectionSelectTypeTransportLabel')}</legend>
                    {this.renderTransportOptions()}
                  </fieldset>
                </MultiChoice>
              </div>
            </GridCol>
          </GridRow>
          <GridRow>
            <SaveAsDraftButton
              formactionUrl={saveAsDraftFormActionUrl}
              onClick={this.onSaveAsDraft}
            />
            <ContinueButton id="continue" type="submit">
              {t('commonContinueButtonSaveAndContinueButton')}
            </ContinueButton>
          </GridRow>
        </form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey} />
      </Main>
    );
  };
}

function mapStateToProps(state) {
  return {
    errors: state.errors,
    transport: state.transport,
    exportLocation: state.exportLocation,
    exportPayload: state.exportPayload,
    storageNotes: state.storageNotes,
    getAddedLandingsType: state.landingsType,
  };
}

function loadData(store, journey) {
  if (Object.keys(this.queryParams).length !== 0) {
    store.dispatch(dispatchApiCallFailed(JSON.parse(this.queryParams.error)));
  }

  return store.dispatch(getTransportDetails(journey, this.documentNumber))
    .then(() => store.dispatch(getExportCountry(this.documentNumber)))
    .then(() => store.dispatch(getStorageNotesFromRedis(this.documentNumber)))
    .then(() => {
      if (journey === catchCertificateJourney) {
        store.dispatch(getLandingType(this.documentNumber));
      }
    });
}

export const component = withTranslation()(connect(
  mapStateToProps,
  {
    addTransport,
    dispatchClearErrors,
    getTransportDetails,
    saveTransport,
    clearTransportDetails,
    getExportCountry,
    getLandingType,
    getFromRedis: getStorageNotesFromRedis
  }
)(TransportSelectionPage));

TransportSelectionPage.propTypes = {
  transport: PropTypes.object.isRequired,
  errors: PropTypes.object,
  exportLocation: PropTypes.object
};

export default {
  loadData,
  component: PageTemplateWrapper(component)
};
