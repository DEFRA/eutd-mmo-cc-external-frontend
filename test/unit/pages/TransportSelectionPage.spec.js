import { mount } from 'enzyme';
import * as React from 'react';
import {component as TransportSelectionPage} from '../../../src/client/pages/common/transport/TransportSelectionPage';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import { ALL_SUPPORTED_VEHICLE_TYPES_NO_DIRECT_LANDING } from '../../../src/client/components/helper/vehicleType';
import {
  addTransport,
  getTransportDetails,
  clearTransportDetails,
  getExportCountry,
  getStorageNotesFromRedis
} from '../../../src/client/actions';
import { getLandingType } from '../../../src/client/actions/landingsType.actions';
import { catchCertificateJourney, storageNoteJourney } from '../../../src/client/helpers/journeyConfiguration';
import { act } from 'react-dom/test-utils';
import * as ErrorUtils from '../../../src/client/pages/utils/errorUtils';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

jest.mock('../../../src/client/actions');
jest.mock('../../../src/client/actions/landingsType.actions');
const mockPush = jest.fn();

describe('TransportSelectionPage', () => {
  const mockSaveTransport= jest.fn();
  const mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');

  const mockStore = configureStore([thunk.withExtraArgument({
    orchestrationApi: {
      get: () => {
        return new Promise((res) => {
          res({});
        });
      },
      post: () => {
        return new Promise((res) => {
          res({});
        });
      }
    }
  })]);

  const addTransportDetails = jest.fn;

  const props = {
    route: {
      path: ':documentNumber/how-does-the-export-leave-the-uk',
      title: 'Create a UK catch certificate for exports',
      previousUri: ':documentNumber/conservationAndManagement',
      vehicleTypes: ALL_SUPPORTED_VEHICLE_TYPES_NO_DIRECT_LANDING,
      saveAsDraftUri: ':documentNumber/create-catch-certificate/catch-certificates',
      journey: catchCertificateJourney,
      truckCmrUri : ':documentNumber/add-transportation-details-truck',
      planeDetailsUri: ':documentNumber/add-transportation-details-plane',
      trainDetailsUri: ':documentNumber/add-transportation-details-train',
      landingsEntryUri: ':documentNumber/landings-entry',
      containerVesselDetailsUri: ':documentNumber/add-transportation-details-container-vessel',
      summaryUri: ':documentNumber/check-your-information',
      progressUri: ':documentNumber/progress'
    },
    match: {
      params: {
        documentNumber: 'documentNumber'
      }
    },
    addTransportDetails : addTransportDetails,
    history: {
      push: mockPush
    },
  };

  const state = {
    errors: {},
    transport: {},
    getTransportDetails: {},
    clearTransportDetails: {},
    landingsType: {
      landingsEntryOption: 'manualEntry',
      generatedByContent: false
    }
  };

  const mountComponent = (p = props, s = state) =>
    mount(
      <Provider store={mockStore({...s})}>
        <MemoryRouter>
          <TransportSelectionPage {...p} />
        </MemoryRouter>
      </Provider>
    );

  beforeEach(() => {
    addTransport.mockReturnValue({ type: 'ADD_TRANSPORT' });
    getTransportDetails.mockReturnValue({ type: 'GET_TRANSPORT_DETAILS' });
    clearTransportDetails.mockReturnValue({ type: 'CLEAR_TRANSPORT_DETAIL' });
    getExportCountry.mockReturnValue({ type: 'GET_EXPORT_PAYLOAD'});
    getStorageNotesFromRedis.mockReturnValue({ type: 'GET_FROM_REDIS' });
    getLandingType.mockReturnValue({ type: 'GET_LANDING_TYPE' });

    jest.spyOn(window, 'scrollTo')
      .mockReturnValue();

    window.scrollTo = jest.fn();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('all journeys', () => {

    let wrapper;

    beforeEach(async () => {
      wrapper = await mountComponent();
    });

    it('should render plane radio button', () => {
      expect(wrapper.find('input#plane').exists()).toBeTruthy();
    });

    it('should render truck radio button', () => {
      expect(wrapper.find('input#truck').exists()).toBeTruthy();
    });

    it('should render train radio button', () => {
      expect(wrapper.find('input#train').exists()).toBeTruthy();
    });

    it('should render containerVessel radio button', () => {
      expect(wrapper.find('input#containerVessel').exists()).toBeTruthy();
    });

    it('should not render directLanding radio button', () => {
      expect(wrapper.find('input#directLanding').exists()).toBeFalsy();
    });

    it('should not render error island when there are no errors', () => {
      expect(wrapper.find('#errorIsland').exists()).toBeFalsy();
    });

    it('should have a for attribute on all input labels', () => {
      expect(wrapper.find('label#label-truck').props()['htmlFor']).toBeDefined();
      expect(wrapper.find('label#label-truck').props()['htmlFor']).toBe('truck');
      expect(wrapper.find('label#label-plane').props()['htmlFor']).toBeDefined();
      expect(wrapper.find('label#label-plane').props()['htmlFor']).toBe('plane');
      expect(wrapper.find('label#label-train').props()['htmlFor']).toBeDefined();
      expect(wrapper.find('label#label-train').props()['htmlFor']).toBe('train');
      expect(wrapper.find('label#label-containerVessel').props()['htmlFor']).toBeDefined();
      expect(wrapper.find('label#label-containerVessel').props()['htmlFor']).toBe('containerVessel');
    });

    it('should render disabled submit button', () => {
      expect(wrapper.find('button[type="submit"]').exists()).toBeTruthy();
    });

    it('should have an id on all inputs', () => {
      expect(wrapper.find('input[id="truck"]').exists()).toBeTruthy();
      expect(wrapper.find('input[id="plane"]').exists()).toBeTruthy();
      expect(wrapper.find('input[id="train"]').exists()).toBeTruthy();
      expect(wrapper.find('input[id="containerVessel"]').exists()).toBeTruthy();
    });

    it('renders Back to Your Progress link with the correct href property', () => {
      expect(wrapper.find('BackToProgressLink')).toBeDefined();
      expect(wrapper.find('BackToProgressLink').find('a').props().href).toBe('/documentNumber/progress');
    });

    it('should call getTransportDetails with correct params', () => {
      expect(getTransportDetails).toHaveBeenCalled();
      expect(getTransportDetails).toHaveBeenCalledWith('catchCertificate', 'documentNumber');
    });

    it('should call getStorageNotesFromRedis with correct params', () => {
      expect(getStorageNotesFromRedis).toHaveBeenCalled();
      expect(getStorageNotesFromRedis).toHaveBeenCalledWith('documentNumber');
    });

    it('should render error island when there are any errors', async () => {
      const state = {
        errors: {
          errors: [
            {
              targetName: 'vehicle',
              text: 'sdTransportSelectionTypeOfTransportErrorNull',
            },
          ],
        },
        transport: {}
      };

      const mountComponent = (p = props, s = state) =>
        mount(
        <Provider store={mockStore({...s})}>
          <MemoryRouter>
            <TransportSelectionPage {...p} />
          </MemoryRouter>
        </Provider>
      );
      wrapper = await mountComponent();

      expect(wrapper.find('#errorIsland').exists()).toBeTruthy();
    });

    it('should catch any errors on submit and scroll to the error island', async () => {
      const instance = wrapper.find('TransportSelectionPage').instance();
      mockSaveTransport.mockRejectedValue('error');

      await act(async () => {
        await instance.onSubmit({preventDefault: jest.fn()});
      });

      expect(mockScrollToErrorIsland).toHaveBeenCalled();
    });

    it('should catch any errors on save as draft and scroll to the error island', async () => {
      const instance = wrapper.find('TransportSelectionPage').instance();
      mockSaveTransport.mockRejectedValue('error');

      await act(async () => {
        await instance.onSaveAsDraft({preventDefault: jest.fn()});
      });

      expect(mockScrollToErrorIsland).toHaveBeenCalled();
    });
  });

  describe('the catch certificate journey', () => {

    it('should call getLandingType', async () => {
      await act(async () => await mountComponent());

      expect(getLandingType).toHaveBeenCalledTimes(1);
      expect(getLandingType).toHaveBeenCalledWith('documentNumber');
    });

    it('should not go anywhere if the export is for a manual entry', async () => {
      await act(async () => await mountComponent());

      expect(mockPush).not.toHaveBeenCalled();
    });

    it('should go to the progress page if the export is for a direct landing', async () => {
      const s = {
        ...state,
        landingsType: {
          landingsEntryOption: 'directLanding',
          generatedByContent: false
        }
      };

      await act(async () => await mountComponent(props, s));

      expect(mockPush).toHaveBeenCalled();
      expect(mockPush).toHaveBeenCalledWith('documentNumber/progress');
    });

    it('should go to the landing entry page if the landing entry option is null', async () => {
      const s = {
        ...state,
        landingsType: {
          landingsEntryOption: null,
          generatedByContent: false
        }
      };

      await act(async () => await mountComponent(props, s));

      const landingEntryUri = props.route.landingsEntryUri.replace(
        ':documentNumber',
        props.match.params.documentNumber
      );

      expect(mockPush).toHaveBeenCalledWith(landingEntryUri);
      expect(mockPush).toHaveBeenCalledTimes(1);
    });

  });

  describe('the storage document journey', () => {

    let wrapper;

    beforeEach(async () => {
      const p = {
        ...props,
        route: {
          ...props.route,
          journey: storageNoteJourney
        }
      };

      wrapper = await mountComponent(p);
    });

    it('should not call getLandingType', () => {
      expect(getLandingType).not.toHaveBeenCalled();
    });
    
    it('renders Back to Your Progress link with the correct href property', () => {
      expect(wrapper.find('BackToProgressLink')).toBeDefined();
      expect(wrapper.find('BackToProgressLink').find('a').props().href).toBe('/documentNumber/progress');
    });
  });

});

describe('TransportSelectionPage snapshot', () => {
  let wrapper;
  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res({});
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);
  const mockPush = jest.fn();
  const store = mockStore({
    errors: {},
    transport: {
      vehicle: 'plane',
    },
    getTransportDetails: {},
    landingsType: {
      generatedByContent: false,
      landingsEntryOption: 'directLanding',
    },
  });
  const props = {
    route: {
      path: ':documentNumber/how-does-the-export-leave-the-uk',
      title: 'Create a UK catch certificate for exports',
      previousUri: ':documentNumber/conservationAndManagement',
      vehicleTypes: ALL_SUPPORTED_VEHICLE_TYPES_NO_DIRECT_LANDING,
      saveAsDraftUri:
        ':documentNumber/create-catch-certificate/catch-certificates',
      journey: catchCertificateJourney,
      truckCmrUri: ':documentNumber/add-transportation-details-truck',
      planeDetailsUri: ':documentNumber/add-transportation-details-plane',
      trainDetailsUri: ':documentNumber/add-transportation-details-train',
      landingsEntryUri: ':documentNumber/landings-entry',
      containerVesselDetailsUri:
        ':documentNumber/add-transportation-details-container-vessel',
      summaryUri: ':documentNumber/check-your-information',
      progressUri: ':documentNumber/progress',
    },
    match: {
      params: {
        documentNumber: 'GBR-23423-4234234',
      },
    },
    history: {
      push: mockPush,
    },
  };
  beforeEach(() => {
    getTransportDetails.mockReturnValue({ type: 'GET_TRANSPORT_DETAILS' });
    getLandingType.mockReset();
    getLandingType.mockReturnValue({ type: 'GET_LANDINGS_TYPE' });
    clearTransportDetails.mockReturnValue({ type: 'clear_transport_data' });
    window.scrollTo = jest.fn();

    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <TransportSelectionPage {...props} />
        </MemoryRouter>
      </Provider>
    );
  });

  afterEach(() => {
    mockPush.mockReset();
  });
  it('should generate snapshot for TransportSelectionPage', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
});

