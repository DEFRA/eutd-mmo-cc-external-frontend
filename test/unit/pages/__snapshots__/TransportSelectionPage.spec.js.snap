// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransportSelectionPage snapshot should generate snapshot for TransportSelectionPage 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="GBR-23423-4234234/conservationAndManagement"
          >
            Back
          </a>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            How does the export leave the UK?
          </h1>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/transport/add"
        method="POST"
      >
        <input
          name="currentUri"
          type="hidden"
          value=":documentNumber/how-does-the-export-leave-the-uk"
        />
        <input
          name="truckCmrUri"
          type="hidden"
          value=":documentNumber/add-transportation-details-truck"
        />
        <input
          name="planeDetailsUri"
          type="hidden"
          value=":documentNumber/add-transportation-details-plane"
        />
        <input
          name="trainDetailsUri"
          type="hidden"
          value=":documentNumber/add-transportation-details-train"
        />
        <input
          name="containerVesselDetailsUri"
          type="hidden"
          value=":documentNumber/add-transportation-details-container-vessel"
        />
        <input
          name="summaryUri"
          type="hidden"
          value=":documentNumber/check-your-information"
        />
        <input
          name="journey"
          type="hidden"
          value="catchCertificate"
        />
        <input
          name="dashboardUri"
          type="hidden"
          value=":documentNumber/create-catch-certificate/catch-certificates"
        />
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-1w80l5h-StyledColumn e4u1u2g0"
          >
            <div
              id="vehicle"
            >
              <div
                class="e1vcssft0 css-1phgi8s-StyledFieldset-StyledHoc e4u1u2g0"
              >
                <span
                  class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
                >
                  Select a type of transport
                </span>
                <fieldset>
                  <legend
                    class="visually-hidden"
                  >
                    Select a type of transport
                  </legend>
                  <label
                    class="css-sdrcg6-StyledHoc"
                    for="truck"
                    id="label-truck"
                  >
                    <input
                      class="css-1xt3i3l"
                      id="truck"
                      name="vehicle"
                      type="radio"
                      value="truck"
                    />
                    <span
                      class="css-1uwxoxa"
                    >
                      Truck
                    </span>
                  </label>
                  <label
                    class="css-sdrcg6-StyledHoc"
                    for="plane"
                    id="label-plane"
                  >
                    <input
                      checked=""
                      class="css-1xt3i3l"
                      id="plane"
                      name="vehicle"
                      type="radio"
                      value="plane"
                    />
                    <span
                      class="css-1uwxoxa"
                    >
                      Plane
                    </span>
                  </label>
                  <label
                    class="css-sdrcg6-StyledHoc"
                    for="train"
                    id="label-train"
                  >
                    <input
                      class="css-1xt3i3l"
                      id="train"
                      name="vehicle"
                      type="radio"
                      value="train"
                    />
                    <span
                      class="css-1uwxoxa"
                    >
                      Train
                    </span>
                  </label>
                  <label
                    class="css-sdrcg6-StyledHoc"
                    for="containerVessel"
                    id="label-containerVessel"
                  >
                    <input
                      class="css-1xt3i3l"
                      id="containerVessel"
                      name="vehicle"
                      type="radio"
                      value="containerVessel"
                    />
                    <span
                      class="css-1uwxoxa"
                    >
                      Container vessel
                    </span>
                  </label>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/transport/add/saveAsDraft"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
            type="submit"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/GBR-23423-4234234/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
